### PR TITLE
Add `RestClient` for Java, Kotlin Coroutine and Scala

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`WebClient` and `WebClientPreparation` has powerful and fluent features.
RESTful API is classic but it is still the dominant protocol for sharing
web resources.
The `WebClient` is able to handle not only REST responses but also
a stream response and a file response.
The API design of `WebClient`, in turn, is not optimized for REST API.
See #4250 

Modifications:

- Add `RestClient` for exchange RESTful APIs.
  - It assume that the response has a JSON body.
- Add `RestClientBuilder` so as to fluently build a `RestClient`.
- Refactor the intefaces for request preparations to accommodate
  `RestClientPreparation`.
  - Method and path setters are extracted to `RequestMethodSetters`.
  - Query and path param setters are moved to `PathAndQueryParamSetters`
  - Added `HttpMessageSetters` for setting `HttpMessage` properties.
- Add `execute()` extension methods for Kotlin Coroutine.
- Add `ScalaRestClient` for Scala.
  - Tried to add extension methods using an implicit class but Scala
    compiler failed to find the overloaded method which has a generic
    parameter.
- Provide a extension method to convert a `WebClient` to
  `ScalaRestClient`.
- Add `restClient()` to `ServerExtension` and `ServerRule`.

Result:

You can now easily send and receive RESTful APIs using `RestClient`
- Java
  ```java
  RestClient restClient = RestClient.of("...");
  CompletableFuture<ResponseEntity<Customer>> response =
      restClient.get("/api/v1/customers/{customerId}")
                .pathParam("customerId", "0000001")
                .execute(Customer.class);
  ```
- Kotlin
  ```kt
  val restClient: RestClient = RestClient.of("...");
  val response: ResponseEntity<Customer> =
     restClient
       .get("/api/v1/customers/{customerId}")
       .pathParam("customerId", "0000001")
       .execute<Customer>()  // a suspend function
  ```
- Scala
  ```scala
  val restClient: ScalaRestClient = ScalaRestClient("...")
  val response: Future[ResponseEntity[Result]] =
    restClient.post("/api/v1/customers")
              .contentJson(new Customer(...))
              .execute[Result]()
  ```

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/HttpRequestSetters.java [hunk 1]
```diff
@@ -29,110 +29,67 @@
...
-    /**
-     * Sets the path for this request.
-     */
-    HttpRequestSetters path(String path);
+public interface HttpRequestSetters extends RequestMethodSetters, PathAndQueryParamSetters, HttpMessageSetters {
 
     /**
      * Sets the content as UTF_8 for this request.
      */
+    @Override
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/HttpRequestSetters.java:37]: I understand what you are concerned about.
But there is a small difference in the Javadoc. It changed `message` to `request`.

File: core/src/main/java/com/linecorp/armeria/client/RequestPreparationSetters.java [hunk 2]
```diff
@@ -18,21 +18,22 @@
...
-interface RequestPreparationSetters<T> extends HttpRequestSetters, RequestOptionsSetters {
-
-    /**
-     * Builds and executes the request.
-     */
-    T execute();
+/**
+ * Sets properties for building an {@link HttpRequest} and {@link RequestOptions}.
+ */
+public interface RequestPreparationSetters extends PathAndQueryParamSetters, HttpMessageSetters,
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/RequestPreparationSetters.java:30]: We need it.

File: core/src/main/java/com/linecorp/armeria/common/RequestMethodSetters.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+/**
+ * Sets properties for building a {@link HttpMethod} and a request path.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/RequestMethodSetters.java:22]: Ditto

**Review by trustin — COMMENTED** [review]
Excellent work, @ikhoon. Left some minor comments 🙇

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 4]
```diff
@@ -598,6 +598,14 @@ default HttpResponse trace(String path, @Nullable QueryParams params) {
     @UnstableApi
     BlockingWebClient blocking();
 
+    /**
+     * Returns a {@link RestClient} that connects to the same {@link URI} with this {@link WebClient}.
+     */
+    @UnstableApi
+    default RestClient toRestClient() {
+        return RestClient.of(this);
+    }
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java:607]: Sound good.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 5]
```diff
@@ -43,6 +48,8 @@ abstract class AbstractHttpMessageBuilder {
     @Nullable
     private HttpHeadersBuilder httpTrailers;
 
+    protected AbstractHttpMessageBuilder() {}
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:54]: I figured out that the default scope of `MissingJavadocMethod` is `public`. https://checkstyle.sourceforge.io/config_javadoc.html#MissingJavadocMethod
Checked the scope to `projected` and fixed other violations together.

File: settings/checkstyle/checkstyle.xml [hunk 6]
```diff
@@ -208,7 +208,9 @@
       <property name="allowMissingReturnTag" value="true"/>
       <property name="allowedAnnotations" value="Override, Test"/>
     </module>
-    <module name="MissingJavadocMethod"/>
+    <module name="MissingJavadocMethod">
+      <property name="scope" value="protected"/>
```
Comment (ikhoon) [inline settings/checkstyle/checkstyle.xml:212]: note to self: `protected` scope covers both `public` and `protected`
https://checkstyle.sourceforge.io/property_types.html#Scope

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:54]: Thanks, @ikhoon !

**Review by trustin — APPROVED** [review]
Nice work, @ikhoon 🚀

File: core/src/main/java/com/linecorp/armeria/client/RequestPreparationSetters.java [hunk 7]
```diff
@@ -18,21 +18,24 @@
...
 import io.netty.util.AttributeKey;
 
-interface RequestPreparationSetters<T> extends HttpRequestSetters, RequestOptionsSetters {
-
-    /**
-     * Builds and executes the request.
-     */
-    T execute();
+/**
+ * provides the setters for building an {@link HttpRequest} and {@link RequestOptions}.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/RequestPreparationSetters.java:29]:
```suggestion
 * Provides the setters for building an {@link HttpRequest} and {@link RequestOptions}.
```

File: core/src/main/java/com/linecorp/armeria/client/RestClientPreparation.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+    /**
+     * Sends the HTTP request and converts the JSON response body as the {@code T} object using the default
+     * {@link ObjectMapper}.
+     *
+     * @see JacksonObjectMapperProvider
+     */
+    public <T> CompletableFuture<ResponseEntity<T>> execute(Class<? extends T> clazz) {
+        requireNonNull(clazz, "clazz");
+        @SuppressWarnings("ThreadJoinLoop")
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/RestClientPreparation.java:67]: It seems like cruft?

File: core/src/main/java/com/linecorp/armeria/client/retry/BackoffWrapper.java [hunk 9]
```diff
@@ -22,6 +22,9 @@
  */
 public class BackoffWrapper extends AbstractUnwrappable<Backoff> implements Backoff {
 
+    /**
+     * Creates an {@link BackoffWrapper} with the specified {@code delegate}.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/retry/BackoffWrapper.java:26]:
```suggestion
     * Creates a {@link BackoffWrapper} with the specified {@code delegate}.
```

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java [hunk 10]
```diff
@@ -285,11 +274,10 @@ final HttpHeadersBuilder headersBuilder() {
 
     private RequestHeaders requestHeaders() {
         requestHeadersBuilder.path(buildPath());
-        if (cookies != null) {
-            requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
-        }
+        @SuppressWarnings("resource")
         final HttpData content = content();
         if (content == null || content.isEmpty()) {
+            //noinspection ReactiveStreamsUnusedPublisher
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:280]: cruft?

File: core/src/main/java/com/linecorp/armeria/common/PathAndQueryParamSetters.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    PathAndQueryParamSetters pathParam(String name, Object value);
+
+    /**
+     * Sets multiple path params for this message.
+     */
+    PathAndQueryParamSetters pathParams(Map<String, ?> pathParams);
+
+    /**
+     * Disables path parameters substitution. If path parameter is not disabled and a parameter's, specified
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/PathAndQueryParamSetters.java:41]:
```suggestion
     * Disables path parameters substitution. If path parameter is not disabled and a parameter is specified
```

File: core/src/main/java/com/linecorp/armeria/common/RequestMethodSetters.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+/**
+ * Provide the setters for building a {@link HttpMethod} and a request path.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/RequestMethodSetters.java:22]:
```suggestion
 * Provide the setters for building an {@link HttpMethod} and a request path.
```

**Review by minwoox — APPROVED** [review]
Left only nits.
Thanks! 👍 👍 👍

File: core/src/main/java/com/linecorp/armeria/client/RestClientPreparation.java (hunk 8 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/RestClientPreparation.java:67]: I don't know why IntelliJ warn about this. But now I checked it again and the warning has gone. 
It seems an IntelliJ bug that was fixed by upgrading. 🤪

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java (hunk 10 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:280]: No, it wasn't. I intentionally added it to suppress the warning. 
image

**Comment by minwoox** [comment]
Thanks for adding this nice feature, @ikhoon!

**Pull request merged**